File: lainexperiment-cpp/src/hackerrank/mozilla/compress-the-string.cpp
/*
 *
 * This source file is a part of lainexperiment project.
 * Description for it can be found in ReadMe.txt.
 *
 */

/*
 *
 * Date: 20/03/2016
 *
 * Hacker rank
 * Mozilla challenge
 * Problem: Compress the string
 * Status: accepted
 *
 * Problem
 *
 * Compress the string using run-length algorithm.
 *
 * Sample
 *
 * Input
aaee
 *
 * Output
a2e2
 *
 */

#include <assert.h>
#include <list>
#include <vector>
#include <stack>
#include <iostream>
#include <sstream>
#include <iterator>

using namespace std;

string compress(string str) {
    if (str.empty()) return "";
    ostringstream res;
    string::const_iterator i = str.begin();
    char prev = *i;
    int cnt = 0;
    while (i != str.end()) {
        char ch = *i++;
        if (prev == ch) {
            cnt++;
            continue;
        }
        res << prev;
        prev = ch;
        if (cnt > 1)
            res << cnt;
        cnt = 1;
    }
    if (cnt > 0)
        res << prev;
    if (cnt > 1)
        res << cnt;
    return res.str();
}

int main() {
    cout << compress("aaee") << endl;
}

File: lainexperiment-java/src/lainexperiment/misc/Compress_the_string.java
/* 
 *
 * This source file is a part of lainexperiment project.
 * Description for it can be found in ReadMe.txt.
 *
 */

package lainexperiment.misc;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * <pre>{@code
 * Date: 07/02/2015
 * 
 * Problem
 * 
 * You are given a string FOOFIGHTERS. You have to come up with an algorithm that 
 * will compress this string.
 *
 * Input
 * 
 * FOOFIGHTERS
 * 
 * Output
 * 
 * F2OFIGHTERS
 * 
 * }</pre>
 */
public class Compress_the_string {

    static int compress(char[] a) {
        int k = 0;
        int i = 0;
        while (i < a.length) {
            if (a[i] != a[k]) {
                a[k++] = a[i++];
                continue;
            }
            while (i < a.length && a[i] == a[k])
                i++;
            if (i - k > 1) {
                a[k] = (char) ('0' + (i - k));
                k++;
            }
            k++;
        }
        return k;
    }

    public static void main(String[] args) {
        char[] a = null;
        int k = 0;
        
        a = "1421".toCharArray();
        k = compress(a);
        assertEquals("1421", String.valueOf(a, 0, k));

        a = "FOOFIGHTERS".toCharArray();
        k = compress(a);
        assertEquals("F2OFIGHTERS", String.valueOf(a, 0, k));

        a = "BWWWWWWWW".toCharArray();
        k = compress(a);
        assertEquals("B8W", String.valueOf(a, 0, k));
    
    }

}
